include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -124,59 +127,60 @@ class pool_memory_resource final : public device_memory_resource {
    *               available in `blocks`.
    */
   block block_from_stream(free_list& blocks,
-                          cudaStream_t blocks_stream,
+                          cudaEvent_t blocks_event,
                           size_t size,
-                          cudaStream_t stream)
+                          cudaStream_t stream,
+                          cudaEvent_t stream_event)
   {
     block const b = blocks.best_fit(size);  // get the best fit block
 
-    // If we found a block associated with a different stream,
-    // we have to synchronize the stream in order to use it
-    if ((blocks_stream != stream) && b.is_valid()) {
-      cudaError_t result = cudaStreamSynchronize(blocks_stream);
-
-      RMM_EXPECTS((result == cudaSuccess ||                    // stream synced
-                   result == cudaErrorInvalidResourceHandle),  // stream deleted
-                  rmm::bad_alloc,
-                  "cudaStreamSynchronize failure");
-
-      // Now that this stream is synced, insert all other blocks into this stream's list
-      // Note: This could cause thrashing between two streams. On the other hand, it reduces
-      // fragmentation by coalescing.
-      stream_free_blocks_[stream].insert(blocks.begin(), blocks.end());
-
-      // remove this stream from the freelist
-      stream_free_blocks_.erase(blocks_stream);
+    if (b.is_valid()) {
+      // If we found a block associated with a different stream, we have to insert a wait on the
+      // stream's associated event into the allocating stream.
+      if (stream_event != blocks_event) {
+        stream_free_blocks_[stream_event].insert(blocks.begin(), blocks.end());
+        stream_free_blocks_.erase(blocks_event);
```

Comment:
An orthogonal optimization is that `stream_free_blocks_.erase(blocks_event);` is searching the `std::map` by key for the map associated with `blocks_event`. This is redundant because we already did the lookup once here: https://github.com/rapidsai/rmm/blob/4110fa0e158e10c1d9db67428771d8dc2d79f6c4/include/rmm/mr/device/pool_memory_resource.hpp#L173

So if you were to instead pass in the _iterator_ to the desired `free_list` to `block_from_stream`, you can access and erase it with only a single lookup into the `std::map` since `std::map::erase` has an overload that takes an iterator to the element to erase.

Reply:
I tried this but since current benchmarks only use one stream there is no benefit. Will leave it out for now.